modules/ximgproc/src/weighted_median_filter.cpp

```diff
@@ -0,0 +1,718 @@
+				else if(balanceWeight < 0)
+				{
+					for(;balanceWeight < 0 && curMedianVal != nI-1; curMedianVal++)
+					{
+						float curWeight = 0;
+						int *nextHist = H[curMedianVal+1];
+						int *nextHf = Hf[curMedianVal+1];
+
+						// Compute weight change by shift cut-point
+						int i=0;
+						do
+						{
+							curWeight += (nextHist[i]<<1)*fPtr[i];
+
+							// Update BCB and maintain the necklace table of BCB
+							updateBCB(BCB[i],BCBf,BCBb,i,nextHist[i]<<1);
+
+							i=nextHf[i];
+						}while(i);
+						balanceWeight += curWeight;
+					}
+				}
+
+				// Weighted median is found and written to the output image
+				if(balanceWeight<0)outImg.ptr<int>(y,x)[0] = curMedianVal+1;
+				else outImg.ptr<int>(y,x)[0] = curMedianVal;
+
+				// Update joint-histogram and BCB when local window is shifted.
+				int fval,gval,*curHist;
+
+				// Add entering pixels into joint-histogram and BCB
+				int rownum = y + r + 1;
+				if(rownum < rows)
+				{
+						int *inputImgPtr = I.ptr<int>(rownum);
+						int *guideImgPtr = F.ptr<int>(rownum);
+						uchar *maskPtr = mask.ptr<uchar>(rownum);
+
+						for(int j=downX;j<=upX;j++)
+						{
+							if(!maskPtr[j])continue;
+
+							fval = inputImgPtr[j];
+							curHist = H[fval];
+							gval = guideImgPtr[j];
+
+							// Maintain necklace table of joint-histogram
+							if(!curHist[gval] && gval)
+							{
+								int *curHf = Hf[fval];
+								int *curHb = Hb[fval];
+
+								int p1=0,p2=curHf[0];
+								curHf[gval]=p2;
+								curHb[gval]=p1;
+								curHf[p1]=curHb[p2]=gval;
+							}
+
+							curHist[gval]++;
+
+							// Maintain necklace table of BCB
+							updateBCB(BCB[gval],BCBf,BCBb,gval,((fval <= medianVal)<<1)-1);
+						}
+					}
+
+
+					// Delete leaving pixels into joint-histogram and BCB
+					rownum = y - r;
+					if(rownum >= 0)
+					{
+						int *inputImgPtr = I.ptr<int>(rownum);
+						int *guideImgPtr = F.ptr<int>(rownum);
+						uchar *maskPtr = mask.ptr<uchar>(rownum);
+
+						for(int j=downX;j<=upX;j++)
+						{
+							if(!maskPtr[j])continue;
+
+							fval = inputImgPtr[j];
+							curHist = H[fval];
+							gval = guideImgPtr[j];
+
+							curHist[gval]--;
+
+							// Maintain necklace table of joint-histogram
+							if(!curHist[gval] && gval)
+							{
+								int *curHf = Hf[fval];
+								int *curHb = Hb[fval];
+
+								int p1=curHb[gval],p2=curHf[gval];
+								curHf[p1]=p2;
+								curHb[p2]=p1;
+							}
+
+							// Maintain necklace table of BCB
+							updateBCB(BCB[gval],BCBf,BCBb,gval,-((fval <= medianVal)<<1)+1);
+						}
+					}
+			}
+		}
+
+		// Deallocate the memory
+		{
+			delete []BCB;
+			delete []BCBf;
+			delete []BCBb;
+			int2D_release(H);
+			int2D_release(Hf);
+			int2D_release(Hb);
+		}
+
+		// end of the function
+		return outImg;
+	}
+}
+
+namespace cv
+{
+namespace ximgproc
+{
+	void weightedMedianFilter(InputArray joint, InputArray src, OutputArray dst, int r, double sigma, std::string weightType, Mat mask)
```

Comment:
Please, replace tabs with 4 spaces and remove additional offset inside namespace:
```.cpp
namespace cv
{
namespace ximgproc
{
void weightedMedianFilter(InputArray joint, InputArray src, OutputArray dst, int r, double sigma, std::string weightType, Mat mask)
{
    CV_Assert(!src.empty());
    ...
```